python/IAM_USER_PERMISSION_BOUNDARY_CHECK/IAM_USER_PERMISSION_BOUNDARY_CHECK.py

```diff
@@ -0,0 +1,466 @@
+# Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the License is located at
+#
+#        http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+
+"""
+#####################################
+##           Gherkin               ##
+#####################################
+
+Rule Name:
+    IAM_USER_PERMISSION_BOUNDARY_CHECK
+
+Description:
+    Check if all the IAM users have permission boundary attached. The rule is NON_COMPLAINT if the permission boundary is not attached to the IAM user.
+
+Trigger:
+    Periodic
+
+Reports on:
+    AWS::IAM::User
+
+Rule Parameters:
+    policyArns (Optional) 
+    Comma-separated list of permission boundary policy ARNs, that are expected to be attached to the IAM Users
+
+Scenarios:
+    Scenario 1:
+    Given: No IAM users in Account
+    Then: Return NOT_APPLICABLE
+
+    Scenario 2:
+    Given: At least 1 IAM User is present in the AWS Account
+    And: No permission boundary policy in the Account
+    Then: Return NOT_APPLICABLE

```

Comment:
Return NON_COMPLIANT

Reply:
It will be NOT_APPLICABLE only as there are no permission boundaries in the account only for the user to be attached with one. We need to check the compliance or not only when users are present and at at-least one policy is set as a permission boundary.

Reply:
Here's the logic. There is at least 1 IAM Users. There is no permission boundary. Therefore no IAM User present in the account has a permission boundary. Hence every single IAM User is NON COMPLIANT. Am I missing something?

Reply:
In the account itself there is no permission boundary present. So it so can happen you select a policy to be permission boundary but never attach it to any user. So this states that there is a permission boundary present in the account but the IAM user does not have it attached. In this case declaring IAM user non- complaint makes sense. But in your account only permission policies are there. You never used permission boundary at all. You never set any policy to be permission boundary for any IAM user at all. Then it so can happen that declaring all IAM users in the account as NON-COMPLAINT may raise a concern to the customer that when I dint set permission boundary at all, like when i dint use it for any user then how is it even considering users for this criteria. Let me know if I am thinking in broader or wrong perspective.

Reply:
> Then it so can happen that declaring all IAM users in the account as NON-COMPLAINT may raise a concern to the customer that when I dint set permission boundary at all, like when i dint use it for any user then how is it even considering users for this criteria.

That the point of the rule, isn't it? If you have any IAM users without permission boundary, then everything is non Compliant.

Reply:
In the account itself there is no permission boundary present. So it so can happen you select a policy to be permission boundary but never attach it to any user. So this states that there is a permission boundary present in the account but the IAM user does not have it attached. In this case declaring IAM user non- complaint makes sense. But in your account only permission policies are there. You never used permission boundary at all. You never set any policy to be permission boundary for any IAM user at all. Then it so can happen that declaring all IAM users in the account as NON-COMPLAINT may raise a concern to the customer that when I dint set permission boundary at all, like when i dint use it for any user then how is it even considering users for this criteria. Let me know if I am thinking in broader or wrong perspective.